packages/vitest/src/types/coverage.ts

```diff
@@ -15,41 +15,89 @@ export type CoverageReporter =
   | 'text-summary'
   | 'text'
 
-export interface C8Options {
+export type CoverageOptions =
+  | NullCoverageOptions & { provider?: null }
+  | C8Options & { provider?: 'c8' }
+  | IstanbulOptions & { provider?: 'istanbul' }
+
+interface BaseCoverageOptions {
   /**
    * Enable coverage, pass `--coverage` to enable
    *
    * @default false
    */
   enabled?: boolean
-  /**
-   * Directory to write coverage report to
-   */
-  reportsDirectory?: string
+
   /**
    * Clean coverage before running tests
    *
    * @default true
    */
   clean?: boolean
+
   /**
    * Clean coverage report on watch rerun
    *
    * @default false
    */
   cleanOnRerun?: boolean
+
   /**
-   * Allow files from outside of your cwd.
-   *
-   * @default false
+   * Directory to write coverage report to
    */
-  allowExternal?: any
+  reportsDirectory?: string
+
   /**
    * Reporters
    *
    * @default 'text'
    */
   reporter?: Arrayable<CoverageReporter>
+}
+
+export interface NullCoverageOptions extends BaseCoverageOptions {
+  enabled: false
+}
+
+export interface IstanbulOptions extends BaseCoverageOptions {
```

Comment:
These `IstanbulOptions` require some API design. Currently these are options of `istanbul-lib-instrument` which already includes some options which are likely not to be configured by the user.

Instead we should be providing a subset of configration options of all used [istanbul packages](https://github.com/istanbuljs/istanbuljs/tree/master/packages). Currently we are using `istanbul-lib-coverage`, `istanbul-lib-instrument`, `istanbul-lib-report`, `istanbul-lib-source-maps` and `istanbul-reports`. For example the `istanbul-lib-report` package contains [`skipFull` and `skipEmpty` flags](https://github.com/istanbuljs/istanbuljs/tree/master/packages/istanbul-lib-report#istanbul-lib-report) which are something users might be interested in.

Before deciding the API I don't think there is much point of writing any documentation. One option is to start with a really minimal API and wait for feature requests. 😁

Reply:
I went through the undocumented configuration options of used `istanbul`-packages and checked options which should be exposed to users. There's now `skipFull`, `ignoreClassMethods` and `watermarks`. Also added a custom `exclude` option. 

`nyc` has a `check-coverage` option which can be used to compare coverage against thresholds: https://github.com/istanbuljs/nyc#coverage-thresholds. However this is not something `istanbul`-packages provide out-of-the-box and requires custom implementation. 

References from nyc and c8:
- https://github.com/istanbuljs/nyc/blob/master/index.js#L480
- https://github.com/bcoe/c8/blob/main/lib/commands/check-coverage.js#L38

I think support for thresholds could be left for later - otherwise this PR will just keep growing.

Reply:
Thresholds are now implemented - it wasn't that big feature after all. 

I think all needed configuration options are now supported.